README.rst
===============================
Python AMT Tools
===============================

.. image:: https://img.shields.io/travis/sdague/amt.svg
        :target: https://travis-ci.org/sdague/amt

.. image:: https://img.shields.io/pypi/v/amt.svg
        :target: https://pypi.python.org/pypi/amt


Tools for interacting with Intel's Active Management Technology

Background
----------

AMT is a light weight hardware control interface put into some Intel
based laptops and desktops as a tool for corporate fleets to manage
hardware. It provides the basics of power control, as well as remote
console via VNC. It functions by having a dedicated service processor
sniff traffic off the network card on specific ports before it gets to
the operating system. Some versions of Intel NUC boxes have AMT, which
make them ideal candidates for building a reasonable cluster in your
basement.

There was once a tool called ``amttool`` which let you interact with
these systems from Linux. This used the SOAP interface to AMT. That
was removed in v9 of the firmware, which means it no longer works with
modern AMT in the field.

The interface that remains is CIM, a standard from the DMTF that
builds XML models for all the things. There exist very few examples
for how to make this work on the internet, with one exception: the
OpenStack Baremetal (Ironic) service. It has native support for AMT
hardware control.

This project is derivative work from Ironic. The heavy lifting of
understanding all the CIM magic incantations, and oh the magic they
are, comes from that code. Refactored for a more minimal usage.

Hardware that includes AMT
--------------------------

AMT is branded as vPro in products by Intel. It is found in many Intel
based laptops. There are also specific models of Intel NUC that
include vPro.

* `Intel NUC KIT Core Processor BLKNUC5I5MYHE <http://amzn.to/1OZshhF>`_

This code gets tested with ``5i5MYHE`` NUCs as well as an older NUC
that I have laying around.


Configuring AMT
---------------

AMT must be enabled in the BIOS before it can be used externally. This
is done by pressing ``Ctrl-P`` during initial boot. Initial user /
pass is ``admin`` / ``admin``. You will be required to create a new
admin password that has at least 1: number, capital letter, and non
alphanumeric symbol.

One you do that, reboot and you are on your way.

amtctrl
-------

The ``amt`` library installs a binary ``amtctrl`` for working with AMT
enabled machines.

machine enrollment
~~~~~~~~~~~~~~~~~~

To simplify the control commands ``amtcrtl`` has a machine
registry. New machines are added via:

    amtctrl add <name> <address> <amtpassword>

You can see a list of all machines with:

   amtctrl list

And remove an existing machine with:

   amtctrl rm <name>


controlling machines
~~~~~~~~~~~~~~~~~~~~

Once machines are controlled you have a number of options exposed:

   amtctrl <name> <command>

Command is one of:

* on - power on the machine

* off - power off the machine

* reboot - power cycle the machine

* pxeboot - set the machine to pxeboot the next time it reboots, and
  reboot the machine. This is extremely useful if you have install
  automation on pxeboot.

* status - return power status as an ugly CIM blob (TODO: make this better)

Futures
-------

* More extensive in tree testing (there currently is very little of
  this)

* Retry http requests when they fail. AMT processors randomly drop
some connections, built in limited retry should be done.

* Fault handling. The current code is *very* optimistic. Hence, the
  0.x nature.

* Remove console control. There are AMT commands to expose a VNC
  remote console on the box. Want to support those.


HISTORY.rst
.. :changelog:

=========
 History
=========

0.8.0 (2017-06-27)
------------------
* add support for hybernate power state (thanks Chen Rotem Levy)
* ensure hostdb is not world readable when created (thanks bodo-riediger)

0.7.0 (2017-03-24)
------------------
* add support for -p flag to prompt for password and bypass hostdb
* nice errors to the user when auth fails

0.6.0 (2016-10-27)
------------------
* add support for stand-by power state (thanks Chen Rotem Levy)

0.5.0 (2016-10-20)
------------------
* python 2 fixes for python 3 issues (thanks Chen Rotem Levy)

0.4.0 (2016-10-10)
------------------
* python 3 fixes and testing (thanks Rixillo)

0.3.0 (2015-10-08)
------------------
* python 3 print fix (thanks Mike Nawrocki)
* add set_next_boot and change_boot_order_request (thanks John L. Villalovos)

0.1.0 (2015-08-09)
---------------------

* First release on PyPI.


tests/test_amt.py
#!/usr/bin/env python

"""
test_amt
----------------------------------

Tests for `amt` module.
"""
import mock
import unittest

from amt import wsman


def fake_uuid4():
    return "00000000-1111-2222-3333-444455556666"


class BaseTestCase(unittest.TestCase):

    def assertXmlEqual(self, one, two):
        one = one.strip()
        two = two.strip()
        array1 = [x.strip() for x in str(one).split("\n")]
        array2 = [x.strip() for x in str(two).split("\n")]
        self.maxDiff = None
        self.assertEqual(array1, array2)


class TestXMLGen(BaseTestCase):

    def setUp(self):
        pass

    @mock.patch('uuid.uuid4', fake_uuid4)
    def test_get_request(self):
        uri = 'http://10.42.0.50:16992/wsman'
        res = ('http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/'
               'CIM_AssociatedPowerManagementService')

        shouldbe = """<?xml version="1.0" encoding="UTF-8"?>
<s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope" xmlns:wsa="http://schemas.xmlsoap.org/ws/2004/08/addressing" xmlns:wsman="http://schemas.dmtf.org/wbem/wsman/1/wsman.xsd">
<s:Header>
<wsa:Action s:mustUnderstand="true">http://schemas.xmlsoap.org/ws/2004/09/transfer/Get</wsa:Action>
<wsa:To s:mustUnderstand="true">http://10.42.0.50:16992/wsman</wsa:To>
<wsman:ResourceURI s:mustUnderstand="true">http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_AssociatedPowerManagementService</wsman:ResourceURI>
<wsa:MessageID s:mustUnderstand="true">uuid:00000000-1111-2222-3333-444455556666</wsa:MessageID>
<wsa:ReplyTo>
<wsa:Address>http://schemas.xmlsoap.org/ws/2004/08/addressing/role/anonymous</wsa:Address>
</wsa:ReplyTo>
</s:Header>
<s:Body/>
</s:Envelope>"""  # noqa

        self.assertXmlEqual(wsman.get_request(uri, res), shouldbe)

    def tearDown(self):
        pass


class TestAmt(unittest.TestCase):

    def setUp(self):
        pass

    def test_something(self):
        pass

    def tearDown(self):
        pass

# Local Variables:
# turn off whitespace mode because xml is *so* long
# eval: (whitespace-mode -1)
# End:


tests/test_wsman.py
#!/usr/bin/env python

"""
test_amt
----------------------------------

Tests for `amt` module's wsman.py file
"""
import mock
import testtools

from amt import wsman


def fake_uuid4():
    return "00000000-1111-2222-3333-444455556666"


class BaseTestCase(testtools.TestCase):

    def assertXmlEqual(self, one, two):
        one = one.strip()
        two = two.strip()
        array1 = [x.strip() for x in str(one).split("\n")]
        array2 = [x.strip() for x in str(two).split("\n")]
        self.maxDiff = None
        self.assertEqual(array1, array2)


class TestXMLGen(BaseTestCase):

    @mock.patch('uuid.uuid4', fake_uuid4)
    def test_get_request(self):
        uri = 'http://10.42.0.50:16992/wsman'
        res = ('http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/'
               'CIM_AssociatedPowerManagementService')

        shouldbe = """<?xml version="1.0" encoding="UTF-8"?>
<s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope" xmlns:wsa="http://schemas.xmlsoap.org/ws/2004/08/addressing" xmlns:wsman="http://schemas.dmtf.org/wbem/wsman/1/wsman.xsd">
<s:Header>
<wsa:Action s:mustUnderstand="true">http://schemas.xmlsoap.org/ws/2004/09/transfer/Get</wsa:Action>
<wsa:To s:mustUnderstand="true">http://10.42.0.50:16992/wsman</wsa:To>
<wsman:ResourceURI s:mustUnderstand="true">http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_AssociatedPowerManagementService</wsman:ResourceURI>
<wsa:MessageID s:mustUnderstand="true">uuid:00000000-1111-2222-3333-444455556666</wsa:MessageID>
<wsa:ReplyTo>
<wsa:Address>http://schemas.xmlsoap.org/ws/2004/08/addressing/role/anonymous</wsa:Address>
</wsa:ReplyTo>
</s:Header>
<s:Body/>
</s:Envelope>"""  # noqa

        self.assertXmlEqual(wsman.get_request(uri, res), shouldbe)

    @mock.patch('uuid.uuid4', fake_uuid4)
    def test_change_boot_to_pxe_request(self):
        uri = 'http://10.42.0.50:16992/wsman'

        shouldbe = """<?xml version="1.0" encoding="UTF-8"?>
<s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope" xmlns:wsa="http://schemas.xmlsoap.org/ws/2004/08/addressing" xmlns:wsman="http://schemas.dmtf.org/wbem/wsman/1/wsman.xsd" xmlns:n1="http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_BootConfigSetting">
<s:Header>
<wsa:Action s:mustUnderstand="true">http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_BootConfigSetting/ChangeBootOrder</wsa:Action>
<wsa:To s:mustUnderstand="true">http://10.42.0.50:16992/wsman</wsa:To>
<wsman:ResourceURI s:mustUnderstand="true">http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_BootConfigSetting</wsman:ResourceURI>
<wsa:MessageID s:mustUnderstand="true">uuid:00000000-1111-2222-3333-444455556666</wsa:MessageID>
<wsa:ReplyTo>
    <wsa:Address>http://schemas.xmlsoap.org/ws/2004/08/addressing/role/anonymous</wsa:Address>
</wsa:ReplyTo>
<wsman:SelectorSet>
   <wsman:Selector Name="InstanceID">Intel(r) AMT: Boot Configuration 0</wsman:Selector>
</wsman:SelectorSet>
</s:Header>
<s:Body>
  <n1:ChangeBootOrder_INPUT>
     <n1:Source>
        <wsa:Address>http://schemas.xmlsoap.org/ws/2004/08/addressing/role/anonymous</wsa:Address>
        <wsa:ReferenceParameters>
            <wsman:ResourceURI>http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_BootSourceSetting</wsman:ResourceURI>
            <wsman:SelectorSet>
                <wsman:Selector wsman:Name="InstanceID">Intel(r) AMT: Force PXE Boot</wsman:Selector>
            </wsman:SelectorSet>
         </wsa:ReferenceParameters>
     </n1:Source>
   </n1:ChangeBootOrder_INPUT>
</s:Body></s:Envelope>"""  # noqa

        self.assertXmlEqual(wsman.change_boot_to_pxe_request(uri), shouldbe)

    def test_change_boot_order_request_invalid_boot_device(self):
        uri = 'http://10.42.0.50:16992/wsman'
        self.assertRaises(AssertionError,
                          wsman.change_boot_order_request,
                          uri, 'pxe2')


class TestFriendlyPowerState(testtools.TestCase):

    def test_power_on(self):
        self.assertEqual(wsman.friendly_power_state('2'), 'on')
        self.assertEqual(wsman.friendly_power_state(2), 'on')

    def test_power_off(self):
        self.assertEqual(wsman.friendly_power_state('8'), 'off')
        self.assertEqual(wsman.friendly_power_state(8), 'off')

    def test_standby(self):
        self.assertEqual(wsman.friendly_power_state('4'), 'standby')
        self.assertEqual(wsman.friendly_power_state(4), 'standby')

    def test_reboot(self):
        self.assertEqual(wsman.friendly_power_state('5'), 'reboot')
        self.assertEqual(wsman.friendly_power_state(5), 'reboot')

    def test_hibernate(self):
        self.assertEqual(wsman.friendly_power_state('7'), 'hibernate')
        self.assertEqual(wsman.friendly_power_state(7), 'hibernate')

    def test_unknown(self):
        self.assertEqual(wsman.friendly_power_state('42'), 'unknown')
        self.assertEqual(wsman.friendly_power_state(42), 'unknown')

    def test_non_integer(self):
        self.assertRaises(ValueError, wsman.friendly_power_state, 'not-a-number')


class TestAmt(testtools.TestCase):

    def test_something(self):
        pass

# Local Variables:
# turn off whitespace mode because xml is *so* long
# eval: (whitespace-mode -1)
# End:


tests/test_hostdb.py
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.

"""
test_hostdb
----------------------------------

Tests for `amt` module's wsman.py file
"""

import appdirs
import fixtures
import mock
import testtools

from amt import hostdb


class TestHostDB(testtools.TestCase):

    def setUp(self):
        super(TestHostDB, self).setUp()
        dirname = fixtures.TempDir()
        self.useFixture(dirname)
        self.out = self.useFixture(fixtures.StringStream('stdout'))
        self.useFixture(fixtures.MonkeyPatch('sys.stdout', self.out.stream))
        with mock.patch('appdirs.user_config_dir', return_value=dirname.path):
            self.db = hostdb.HostDB()

    @property
    def stdout(self):
        return self.out._details['stdout'].as_text()

    def test_list_servers(self):
        self.db.add_server("os1", "10.42.0.50", "foo")
        self.db.add_server("os2", "10.42.0.51", "foo")
        self.db.list_servers()
        self.assertEqual(self.stdout,
                        "Available servers (2):\n    os1\n    os2")

    def test_rm_servers(self):
        self.db.add_server("os1", "10.42.0.50", "foo")
        self.db.add_server("os2", "10.42.0.51", "foo")
        self.db.add_server("os3", "10.42.0.52", "foo")
        self.db.rm_server("os2")
        self.db.list_servers()
        self.assertEqual(self.stdout,
                        "Available servers (2):\n    os1\n    os3")

    def test_get_server(self):
        self.db.add_server("os1", "10.42.0.50", "foo")
        self.db.add_server("os2", "10.42.0.51", "foo")
        server = self.db.get_server("os1")
        self.assertEqual(server,
                         dict(host="10.42.0.50", passwd="foo"))


amt/amt.py
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.
#
# This file is heavily derived from the Ironic AMT driver at
# https://github.com/openstack/ironic/tree/master/ironic/drivers/modules/amt
#
# Thanks much for the hard work of people in that project to produce
# Open Source software that acts as one of the few bits of example
# code for this interface.

import xml.dom.minidom
from xml.etree import ElementTree

import requests
from requests.auth import HTTPDigestAuth

import wsman


"""CIM schema urls

Conceptually you can query a Service, everything else is for update
only or modeling only. And, yes this is as redundant as it looks.
"""

SCHEMA_BASE = 'http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/'

CIM_AssociatedPowerManagementService = (SCHEMA_BASE +
                                        'CIM_AssociatedPowerManagementService')
CIM_PowerManagementService = (SCHEMA_BASE +
                              'CIM_PowerManagementService')
CIM_BootService = SCHEMA_BASE + 'CIM_BootService'

CIM_ComputerSystem = SCHEMA_BASE + 'CIM_ComputerSystem'
CIM_BootConfigSetting = SCHEMA_BASE + 'CIM_BootConfigSetting'
CIM_BootSourceSetting = SCHEMA_BASE + 'CIM_BootSourceSetting'

# Additional useful constants
_SOAP_ENVELOPE = 'http://www.w3.org/2003/05/soap-envelope'
_ADDRESS = 'http://schemas.xmlsoap.org/ws/2004/08/addressing'
_ANONYMOUS = 'http://schemas.xmlsoap.org/ws/2004/08/addressing/role/anonymous'
_WSMAN = 'http://schemas.dmtf.org/wbem/wsman/1/wsman.xsd'


# magic ports to connect to
AMT_PROTOCOL_PORT_MAP = {
    'http': 16992,
    'https': 16993,
}


def pp_xml(body):
    """Pretty print format some XML so it's readable."""
    pretty = xml.dom.minidom.parseString(body)
    return pretty.toprettyxml(indent="  ")


class Client(object):
    """AMT client.

    Manage interactions with AMT host.
    """
    def __init__(self, address, password, username='admin', protocol='http'):
        port = AMT_PROTOCOL_PORT_MAP[protocol]
        path = '/wsman'
        self.uri = "%(protocol)s://%(address)s:%(port)s%(path)s" % {
            'address': address,
            'protocol': protocol,
            'port': port,
            'path': path}
        self.username = username
        self.password = password

    def power_on(self):
        """Power on the box."""
        payload = wsman.power_state_request(self.uri, "on")
        resp = requests.post(self.uri,
                             auth=HTTPDigestAuth(self.username, self.password),
                             data=payload)
        if resp.status_code == 200:
            rv = _return_value(resp.content, CIM_PowerManagementService)
            if rv == 0:
                return 0
            print(pp_xml(resp.content))
        else:
            print(resp.content)
            return 1

    def power_off(self):
        """Power off the box."""
        payload = wsman.power_state_request(self.uri, "off")
        resp = requests.post(self.uri,
                             auth=HTTPDigestAuth(self.username, self.password),
                             data=payload)
        if resp.status_code == 200:
            rv = _return_value(resp.content, CIM_PowerManagementService)
            if rv == 0:
                return 0
            print(pp_xml(resp.content))
        else:
            print(resp.content)
            return 1

    def power_cycle(self):
        """Power cycle the box."""
        payload = wsman.power_state_request(self.uri, "reboot")
        resp = requests.post(self.uri,
                             auth=HTTPDigestAuth(self.username, self.password),
                             data=payload)
        if resp.status_code == 200:
            rv = _return_value(resp.content, CIM_PowerManagementService)
            if rv == 0:
                return 0
            print(pp_xml(resp.content))
        else:
            print(resp.content)
            return 1

    def pxe_next_boot(self):
        """Sets the machine to PXE boot on it's next reboot

        Will default back to normal boot list on the reboot that follows.
        """
        payload = wsman.change_boot_to_pxe_request(self.uri)
        resp = requests.post(self.uri,
                             auth=HTTPDigestAuth(self.username, self.password),
                             data=payload)
        if resp.status_code == 200:
            print(pp_xml(resp.content))
            # rv = _return_value(resp.content, CIM_PowerManagementService)
            # if rv != 0:
            #     print(pp_xml(resp.content))
            #     return 1

        payload = wsman.enable_boot_config_request(self.uri)
        resp = requests.post(self.uri,
                             auth=HTTPDigestAuth(self.username, self.password),
                             data=payload)
        if resp.status_code == 200:
            print(pp_xml(resp.content))
            # rv = _return_value(resp.content, CIM_PowerManagementService)
            # if rv != 0:
            #     print(pp_xml(resp.content))
            #     return 1

    def power_status(self):
        payload = wsman.get_request(
            self.uri,
            CIM_AssociatedPowerManagementService)
        resp = requests.post(self.uri,
                             auth=HTTPDigestAuth(self.username, self.password),
                             data=payload)
        return pp_xml(resp.content)


def _return_value(content, ns):
    """Find the return value in a CIM response.

    The xmlns is needed because everything in CIM is a million levels
    of namespace indirection.
    """
    doc = ElementTree.fromstring(content)
    query = './/{%(ns)s}%(item)s' % {'ns': ns, 'item': 'ReturnValue'}
    rv = doc.find(query)
    return int(rv.text)


amt/hostdb.py
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.

from six.moves import configparser
import os

import appdirs

appauthor = "sdague"
appname = "amtctrl"


class HostDB(object):
    def __init__(self):
        self.confdir = appdirs.user_config_dir(appname, appauthor)
        self.confname = os.path.join(self.confdir, 'hosts.cfg')
        self.config = configparser.ConfigParser()
        self.config.read(self.confname)

    def list_servers(self):
        print("Available servers (%d):" % len(self.config.sections()))
        for item in self.config.sections():
            print("    %s" % item)

    def add_server(self, name, host, passwd):
        self.config.add_section(name)
        self.config.set(name, 'host', host)
        self.config.set(name, 'passwd', passwd)
        # ensure the directory exists
        if not os.path.exists(self.confdir):
            os.makedirs(self.confdir, 0o770)

        with open(self.confname, 'w') as f:
            self.config.write(f)

    def rm_server(self, name):
        self.config.remove_section(name)

        with open(self.confname, 'w') as f:
            self.config.write(f)

    def get_server(self, name):
        if self.config.has_section(name):
            return {
                'host': self.config.get(name, 'host'),
                'passwd': self.config.get(name, 'passwd')
            }
        else:
            print("No config found for server (%s), "
                  "perhaps you need to add one via ``amtctrl add``")


amt/client.py
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.
#
# This file is heavily derived from the Ironic AMT driver at
# https://github.com/openstack/ironic/tree/master/ironic/drivers/modules/amt
#
# Thanks much for the hard work of people in that project to produce
# Open Source software that acts as one of the few bits of example
# code for this interface.

import xml.dom.minidom
from xml.etree import ElementTree

import requests
from requests.auth import HTTPDigestAuth

import amt.wsman


"""CIM schema urls

Conceptually you can query a Service, everything else is for update
only or modeling only. And, yes this is as redundant as it looks.
"""

SCHEMA_BASE = 'http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/'

CIM_AssociatedPowerManagementService = (SCHEMA_BASE +
                                        'CIM_AssociatedPowerManagementService')
CIM_PowerManagementService = (SCHEMA_BASE +
                              'CIM_PowerManagementService')
CIM_BootService = SCHEMA_BASE + 'CIM_BootService'

CIM_ComputerSystem = SCHEMA_BASE + 'CIM_ComputerSystem'
CIM_BootConfigSetting = SCHEMA_BASE + 'CIM_BootConfigSetting'
CIM_BootSourceSetting = SCHEMA_BASE + 'CIM_BootSourceSetting'

# Additional useful constants
_SOAP_ENVELOPE = 'http://www.w3.org/2003/05/soap-envelope'
_ADDRESS = 'http://schemas.xmlsoap.org/ws/2004/08/addressing'
_ANONYMOUS = 'http://schemas.xmlsoap.org/ws/2004/08/addressing/role/anonymous'
_WSMAN = 'http://schemas.dmtf.org/wbem/wsman/1/wsman.xsd'


# magic ports to connect to
AMT_PROTOCOL_PORT_MAP = {
    'http': 16992,
    'https': 16993,
}


def pp_xml(body):
    """Pretty print format some XML so it's readable."""
    pretty = xml.dom.minidom.parseString(body)
    return pretty.toprettyxml(indent="  ")


class Client(object):
    """AMT client.

    Manage interactions with AMT host.
    """
    def __init__(self, address, password, username='admin', protocol='http'):
        port = AMT_PROTOCOL_PORT_MAP[protocol]
        path = '/wsman'
        self.uri = "%(protocol)s://%(address)s:%(port)s%(path)s" % {
            'address': address,
            'protocol': protocol,
            'port': port,
            'path': path}
        self.username = username
        self.password = password

    def post(self, payload, ns=None):
        resp = requests.post(self.uri,
                             headers={'content-type':
                                      'application/soap+xml;charset=UTF-8'},
                             auth=HTTPDigestAuth(self.username, self.password),
                             data=payload)
        resp.raise_for_status()
        if ns:
            rv = _return_value(resp.content, ns)
            if rv == 0:
                return 0
            print(pp_xml(resp.content))
        else:
            return 0

    def power_on(self):
        """Power on the box."""
        payload = amt.wsman.power_state_request(self.uri, "on")
        return self.post(payload, CIM_PowerManagementService)

    def power_off(self):
        """Power off the box."""
        payload = amt.wsman.power_state_request(self.uri, "off")
        return self.post(payload, CIM_PowerManagementService)

    def power_cycle(self):
        """Power cycle the box."""
        payload = amt.wsman.power_state_request(self.uri, "reboot")
        return self.post(payload, CIM_PowerManagementService)

    def pxe_next_boot(self):
        """Sets the machine to PXE boot on its next reboot

        Will default back to normal boot list on the reboot that follows.
        """
        self.set_next_boot(boot_device='pxe')

    def set_next_boot(self, boot_device):
        """Sets the machine to boot to boot_device on its next reboot

        Will default back to normal boot list on the reboot that follows.
        """
        payload = amt.wsman.change_boot_order_request(self.uri, boot_device)
        self.post(payload)

        payload = amt.wsman.enable_boot_config_request(self.uri)
        self.post(payload)

    def power_status(self):
        payload = amt.wsman.get_request(
            self.uri,
            CIM_AssociatedPowerManagementService)
        resp = requests.post(self.uri,
                             auth=HTTPDigestAuth(self.username, self.password),
                             data=payload)
        resp.raise_for_status()
        value = _find_value(
            resp.content,
            CIM_AssociatedPowerManagementService,
            "PowerState")
        return value

    def enable_vnc(self):
        payload = amt.wsman.enable_remote_kvm(self.uri, self.password)
        self.post(payload)
        payload = amt.wsman.kvm_redirect(self.uri)
        self.post(payload)

    def vnc_status(self):
        payload = amt.wsman.get_request(
            self.uri,
            ('http://intel.com/wbem/wscim/1/ips-schema/1/'
             'IPS_KVMRedirectionSettingData'))
        resp = requests.post(self.uri,
                             auth=HTTPDigestAuth(self.username, self.password),
                             data=payload)
        resp.raise_for_status()
        return pp_xml(resp.content)


def _find_value(content, ns, key):
    """Find the return value in a CIM response.

    The xmlns is needed because everything in CIM is a million levels
    of namespace indirection.
    """
    doc = ElementTree.fromstring(content)
    query = './/{%(ns)s}%(item)s' % {'ns': ns, 'item': key}
    rv = doc.find(query)
    return rv.text


def _return_value(content, ns):
    """Find the return value in a CIM response.

    The xmlns is needed because everything in CIM is a million levels
    of namespace indirection.
    """
    doc = ElementTree.fromstring(content)
    query = './/{%(ns)s}%(item)s' % {'ns': ns, 'item': 'ReturnValue'}
    rv = doc.find(query)
    return int(rv.text)
